src/streamed_orderbook.ts

```diff
@@ -0,0 +1,267 @@
+import assert from 'assert'
+import Web3 from 'web3'
+import { EventData } from 'web3-eth-contract'
+import { abi } from '@gnosis.pm/dex-contracts/build/contracts/BatchExchange.json'
+import { BatchExchange } from '@gnosis.pm/dex-contracts/build/types/BatchExchange'
+
+const STARTING_BLOCK = 9340147
+const BLOCK_PAGE_SIZE = 10000
+const BATCH_DURATION = 300
+
+interface AccountsState {
+  accounts: { [key: string]: AccountState },
+  tokens: { [key: string]: string },
+  prices: { [key: string]: bigint },
+}
+
+interface AccountState {
+  balances: { [key: string]: bigint },
+  withdrawals: { [key: string]: bigint },
+  orders: (OrderState | null)[],
+}
+
+interface OrderState {
+  buyToken: string,
+  sellToken: string,
+  validFrom: number,
+  validUntil: number | null,
+  priceNumerator: bigint,
+  priceDenominator: bigint,
+  remainingAmount: bigint,
+}
+
+class StreamedOrderbook {
+  private initialized: boolean = false;
+  private blockNumber: number = 0;
+  private blockBatches: { [key: number]: number } = {};
+  private pendingEvents: any[] = [];
+  private state: AccountsState = {
+    accounts: {},
+    tokens: {},
+    prices: {},
+  };
+
+  constructor(
+    private readonly web3: Web3,
+    private readonly contract: BatchExchange,
+  ) {}
+
+  async init() {
+    if (this.initialized) {
+      return
+    }
+    this.initialized = true
+
+    this.blockNumber = await this.web3.eth.getBlockNumber()
+    for (
+      let fromBlock = STARTING_BLOCK;
+      fromBlock < this.blockNumber;
+      fromBlock += BLOCK_PAGE_SIZE
+    ) {
+      const toBlock = Math.min(fromBlock + BLOCK_PAGE_SIZE - 1, this.blockNumber)
+      console.error(`fetching page ${fromBlock}-${toBlock}`)
+
+      const events = await this.contract.getPastEvents(
+        'allEvents',
+        { fromBlock, toBlock },
+      )
+
+      this.pendingEvents.push(...events)
+      await this.applyAllEventsUntilNextBatch()
+    }
+  }
+
+  dump() {
+    for (const user in this.state.accounts) {
+      console.log(user, this.state.accounts[user])
+    }
+  }
+
+  private async applyAllEventsUntilNextBatch() {
+    if (this.pendingEvents.length === 0) {
+      return
+    }
+
+    const pending = this.pendingEvents.length
+    const nextBatch = await this.eventBatch(this.pendingEvents[pending - 1])
+    for (let i = pending - 2; i >= 0; i--) {
+      const batch = await this.eventBatch(this.pendingEvents[i])
+      if (nextBatch > batch) {
+        this.applyEvents(this.pendingEvents.splice(0, i + 1))
+        return
+      }
+    }
+
+    return false
+  }
+
+  private applyEvents(events: EventData[]) {
```

Comment:
nit: can we reflect in the method name that all events are from the same batch (and maybe assert that in the for loop to make this assumption explicit)?

Reply:
I originally had that, but that required fetching block headers for a bunch of events and made it be too slow.

I will rename the method, but in refactoring I think I can removing this limitation.